### PR TITLE
PCHR-3010: Hotfix Job Contract Absence Types amounts and order

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

Absence Types listed in Job Contract and in Job Contract Modal have wrong order and leave amount values. Also the default entitlements values are not set by default and the "Add holiday" radio button switch is not set by default either. This PR fixes these issues.

## Before

image

## After

**Existing job contract:**

image

**New job contract for new contact:**

image

#### Step 1. Map data correctly regardless of data order and entries amount in data sets:

```js
_.each($scope.leave, function (leaveType) {
  _.extend(leaveType, newScope.leave
  ? _.find(newScope.leave, { leave_type: leaveType.leave_type }) || // Do mapping
  contractRevisionIdObj : contractRevisionIdObj);
});
```

#### Step 2. Fetch data in correct order

```js
function getAbsenceTypes () {
  return AbsenceType.all({ options: { sort: 'weight ASC' } }) // Sort by weight
  .then(AbsenceType.loadCalculationUnits)
  .then(function (absenceTypes) {
    // We do not need indexing anymore since there is no more access
    // to collection entries by their key
    return absenceTypes;
  });
}
```

------------
✅ Manual Tests - passed
✅ Jasmine Tests - passed
✅ JS distributive files - included
⏹ CSS distributive files - not needed
⏹ Backstop Tests - omitted
⏹ PHP Unit Tests - not needed

### CODE REVIEW

File: hrjobcontract/js/src/job-contract/controllers/contract.js [hunk 1]
```diff
@@ -69,8 +69,10 @@ define([
         _.extend($scope.health, newScope.health || contractRevisionIdObj);
         _.extend($scope.pension, newScope.pension || contractRevisionIdObj);
 
-        _.each($scope.leave, function (leaveType, leaveTypeId) {
-          _.extend(leaveType, newScope.leave ? newScope.leave[leaveTypeId] || contractRevisionIdObj : contractRevisionIdObj);
+        _.each($scope.leave, function (leaveType) {
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/controllers/contract.js]: put this in a function and document it, cause it's hard to understand what's going on

File: hrjobcontract/js/src/job-contract/controllers/contract.js [hunk 2]
```diff
@@ -69,8 +69,10 @@ define([
...
+          _.extend(leaveType, newScope.leave
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/controllers/contract.js]: put this whole block
```js
newScope.leave ? _.find(newScope.leave, { leave_type: leaveType.leave_type }) || contractRevisionIdObj : contractRevisionIdObj
```
in a separate function and document it

File: hrjobcontract/js/src/job-contract/controllers/contract.js [hunk 3]
```diff
@@ -186,8 +188,10 @@ define([
           _.extend(entity.health, revisionDetails.health);
           _.extend(entity.pay, revisionDetails.pay);
           _.extend(entity.pension, revisionDetails.pension);
-          _.each(entity.leave, function (leaveType, leaveTypeId) {
-            _.extend(leaveType, revisionDetails.leave ? revisionDetails.leave[leaveTypeId] : '');
+          _.each(entity.leave, function (leaveType) {
+            _.extend(leaveType, revisionDetails.leave
+              ? _.find(revisionDetails.leave, { leave_type: leaveType.leave_type })
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/controllers/contract.js]: how different is this compared to 
```js
newScope.leave ? _.find(newScope.leave, { leave_type: leaveType.leave_type }) || contractRevisionIdObj : contractRevisionIdObj
```
isn't it basically the same except for the fallback value, here it's `''` and before it's `contractRevisionIdObj`

File: hrjobcontract/js/src/job-contract/services/contract-leave.js [hunk 4]
```diff
@@ -35,10 +35,10 @@ define([
        * @return {Promise} resolved with absence types
        */
       function getAbsenceTypes () {
-        return AbsenceType.all({ options: { sort: 'id ASC' } })
+        return AbsenceType.all({ options: { sort: 'weight ASC' } })
         .then(AbsenceType.loadCalculationUnits)
         .then(function (absenceTypes) {
-          return _.indexBy(absenceTypes, 'id');
+          return absenceTypes;
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/services/contract-leave.js]: at this point there is no need for this at all, since you're simply returning the same value

File: hrjobcontract/js/src/job-contract/services/contract-leave.js [hunk 5]
```diff
@@ -210,13 +210,14 @@ define([
...
-            _.each(leaveType, function (type, typeId) {
+            _.each(leaveType, function (type) {
               modelEntry.leave_type = type.id;
               modelEntry.leave_type_title = type.title;
               modelEntry.leave_calculation_unit_name = type.calculation_unit_name;
               modelEntry.leave_calculation_unit_label = type.calculation_unit_label;
-              modelEntry.leave_amount = 0;
+              modelEntry.leave_amount = parseFloat(type.default_entitlement);
               modelEntry.is_add_public_holidays_readonly = type.calculation_unit_name === 'hours';
+              modelEntry.add_public_holidays = !!parseInt(type.add_public_holiday_to_entitlement);
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/services/contract-leave.js]: i don't think `parseInt` is necessary, you're casting to boolean anyway

File: hrjobcontract/js/src/job-contract/services/contract-leave.js [hunk 6]
```diff
@@ -210,13 +210,14 @@ define([
...
             };
 
-            _.each(leaveType, function (type, typeId) {
+            _.each(leaveType, function (type) {
               modelEntry.leave_type = type.id;
               modelEntry.leave_type_title = type.title;
               modelEntry.leave_calculation_unit_name = type.calculation_unit_name;
               modelEntry.leave_calculation_unit_label = type.calculation_unit_label;
-              modelEntry.leave_amount = 0;
+              modelEntry.leave_amount = parseFloat(type.default_entitlement);
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/services/contract-leave.js]: this changes looks weird. Why was it 0 before? There must have been a reason. And is `parseFloat` necessary at all?

File: hrjobcontract/js/test/controllers/contract_test.js [hunk 7]
```diff
@@ -8,6 +8,7 @@ define([
   'job-contract/app'
 ], function (_, moment, contractMock) {
   'use strict';
+  var controller;
```
Comment (AkA84) [inline hrjobcontract/js/test/controllers/contract_test.js]: put this in the first `describe`, where you can see there are all the other "global" vars

File: hrjobcontract/js/test/controllers/contract_test.js [hunk 8]
```diff
@@ -53,6 +54,43 @@ define([
       it('retrieves contract full details', function () {
         expect(ContractService.fullDetails).toHaveBeenCalled();
       });
+
+      describe('when loads the leave data', function () {
```
Comment (AkA84) [inline hrjobcontract/js/test/controllers/contract_test.js]: when *it* loads

File: hrjobcontract/js/test/controllers/contract_test.js [hunk 9]
```diff
@@ -53,6 +54,43 @@ define([
...
+
+        beforeEach(function () {
+          $scope.leave = [{ leave_type: leave.leave_type }];
+          $scope.revisionCurrent.jobcontract_id = 1;
+          $scope.$emit('updateContractView');
+          $scope.$digest();
+          $httpBackend.flush();
+        });
+
+        it('maps Absence Types and Job Contract data correctly', function () {
```
Comment (AkA84) [inline hrjobcontract/js/test/controllers/contract_test.js]: shouldn't you cover all the edge cases:

* when the absence types order don't match the contract data order
* when there are more absence types than contract data (ie some absence type had been added afterwards)
* when there are less absence types than contract data (ie some absence type had been removed/disabled afterwards)

File: hrjobcontract/js/test/services/contract-leave_test.js [hunk 10]
```diff
@@ -44,5 +46,50 @@ define([
...
       });
     });
+
+    describe('model()', function () {
+      var result;
+      var leaveMocks = _.cloneDeep(MockContract.contractLeaves.values);
+      var randomLeaveIndex = _.random(0, leaveMocks.length - 1);
+      var leaveMock = leaveMocks[randomLeaveIndex];
+
+
```
Comment (AkA84) [inline hrjobcontract/js/test/services/contract-leave_test.js]: empty line

File: hrjobcontract/js/src/job-contract/controllers/contract.js (hunk 1 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/controllers/contract.js]: done.

File: hrjobcontract/js/src/job-contract/controllers/contract.js (hunk 2 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/controllers/contract.js]: done.

File: hrjobcontract/js/src/job-contract/controllers/contract.js (hunk 3 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/controllers/contract.js]: I think it is dangerous to amend the existing code at the moment. I try to modify as less as possible now.

File: hrjobcontract/js/src/job-contract/services/contract-leave.js (hunk 4 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/services/contract-leave.js]: Done.

File: hrjobcontract/js/src/job-contract/services/contract-leave.js (hunk 5 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/services/contract-leave.js]: The `type.add_public_holiday_to_entitlement` is a string containing either 0 or 1.

image

File: hrjobcontract/js/test/controllers/contract_test.js (hunk 7 above)
Comment (igorpavlov) [reply hrjobcontract/js/test/controllers/contract_test.js]: Done.

File: hrjobcontract/js/test/controllers/contract_test.js (hunk 8 above)
Comment (igorpavlov) [reply hrjobcontract/js/test/controllers/contract_test.js]: Done.

File: hrjobcontract/js/test/controllers/contract_test.js (hunk 9 above)
Comment (igorpavlov) [reply hrjobcontract/js/test/controllers/contract_test.js]: Done.

File: hrjobcontract/js/test/services/contract-leave_test.js (hunk 10 above)
Comment (igorpavlov) [reply hrjobcontract/js/test/services/contract-leave_test.js]: Done.

File: hrjobcontract/js/src/job-contract/services/contract-leave.js (hunk 6 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/services/contract-leave.js]: You are right, this is unnecessary change. Reverted.

File: hrjobcontract/js/src/job-contract/controllers/contract.js [hunk 11]
```diff
@@ -39,6 +39,35 @@ define([
...
+        });
+      }
+
+      /**
+       * Maps Absence Types with the leave data from the Revision
+       *
+       * @param {Array} leaveEntity
+       * @param {Array} leaveData
+       */
+      function mapLeaveEntityWithRevisionLeaveData (leaveEntity, leaveData) {
```
Comment (AkA84) [inline hrjobcontract/js/src/job-contract/controllers/contract.js]: These two functions are basically the same (I'm really sorry for the placeholder names, but I'm not sure about their exact meaning):

```js
function foo (bar, baz, bax) {
  bax = bax || '';

  _.each(bar, function (leaveType) {
    _.extend(leaveType, baz
      ? _.find(baz, { leave_type: leaveType.leave_type }) || bax
      : bax);
  });
}
```

File: hrjobcontract/js/test/controllers/contract_test.js [hunk 12]
```diff
@@ -53,6 +53,57 @@ define([
...
+
+              $scope.$emit('updateContractView');
+              $scope.$digest();
+              $httpBackend.flush();
+            });
+
+            it('maps Absence Types and Job Contract data correctly', function () {
+              _.each($scope.leave, function (entry) {
+                expect(entry.leave_amount).toEqual(
+                  _.find(leaveMocks, { leave_type: entry.leave_type }).leave_amount);
```
Comment (AkA84) [inline hrjobcontract/js/test/controllers/contract_test.js]:
```js
it('maps Absence Types and Job Contract data correctly', function () {
  _.each($scope.leave, function (entry) {
    expect(entry.leave_amount).toEqual(_.find(leaveMocks, {
      leave_type: entry.leave_type
    }).leave_amount);
  });
});
```

File: hrjobcontract/js/src/job-contract/controllers/contract.js (hunk 11 above)
Comment (igorpavlov) [reply hrjobcontract/js/src/job-contract/controllers/contract.js]: Done.

File: hrjobcontract/js/test/controllers/contract_test.js (hunk 12 above)
Comment (igorpavlov) [reply hrjobcontract/js/test/controllers/contract_test.js]: Done.
